### PR TITLE
WIP: core:(assembly-format) allow empty region blocks to be optional in assembly format

### PR DESCRIPTION
**Pull request by dk949**
This allows non-optional regions to be anchorable by treating missing regions as empty.

As far as I can tell this is consistent with upstream.

### CODE REVIEW

**Comment by dk949** [comment]
I'm not very familiar with this part of the codebase so would really appreciate any tips on what I may have missed.

(will add tests in a little bit)

File: xdsl/irdl/declarative_assembly_format.py [hunk 1]
```diff
@@ -930,6 +930,12 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
         state.last_was_punctuation = False
         state.should_emit_space = True
 
+    def is_anchorable(self) -> bool:
+        return True
```
Comment (alexarice) [inline xdsl/irdl/declarative_assembly_format.py:934]: When setting `is_anchorable` you need to also implement `is_present` which the optional group uses to determine whether to print.

**Review by alexarice — COMMENTED** [review]
This seems like a good change if it agrees with what MLIR does. It would be nice to get a simple test of a region being used as an anchor in `tests/irdl/test_declarative_assembly_format.py`

File: tests/irdl/test_declarative_assembly_format.py [hunk 2]
```diff
@@ -2140,6 +2140,40 @@ class OptionalRegionOp(IRDLOperation):
...
+        ),
+    ],
+)
+def test_optional_groups_empty_regions(format: str, program: str, generic_program: str):
+    """Test the parsing of empty regions in an optional group"""
+
+    @irdl_op_definition
+    class EmptyRegionOp(IRDLOperation):
+        name = "test.empty_region_group"
+        irdl_options = [AttrSizedRegionSegments]
```
Comment (alexarice) [inline tests/irdl/test_declarative_assembly_format.py]: Why do we need this option?

File: tests/irdl/test_declarative_assembly_format.py [hunk 3]
```diff
@@ -2140,6 +2140,40 @@ class OptionalRegionOp(IRDLOperation):
...
+@pytest.mark.parametrize(
+    "format, program, generic_program",
+    [
+        (
+            "(`keyword` $maybe_empty^)? attr-dict",
+            "test.empty_region_group",
+            '"test.empty_region_group"() ({}) : () -> ()',
+        ),
+        (
+            "(`keyword` $maybe_empty^)? attr-dict",
```
Comment (alexarice) [inline tests/irdl/test_declarative_assembly_format.py]: Could we inline the format? it's the same for both.

**Comment by dk949** [comment]
> This seems like a good change if it agrees with what MLIR does. It would be nice to get a simple test of a region being used as an anchor in `tests/irdl/test_declarative_assembly_format.py`

The one thing I'm currently unsure of is whether an anchored region is allowed to be the first directive in an optional group. The way I have it now, it's not

File: tests/irdl/test_declarative_assembly_format.py [hunk 4]
```diff
@@ -2140,6 +2140,40 @@ class OptionalRegionOp(IRDLOperation):
...
+            'test.empty_region_group keyword {\n  "test.op"() : () -> ()\n}',
+            '"test.empty_region_group"() ({ "test.op"() : () -> ()}) : () -> ()',
+        ),
+    ],
```
Comment (alexarice) [inline tests/irdl/test_declarative_assembly_format.py:2158]: Could we add one test with a single empty block? Occured to me that I'm not sure how MLIR treats this.

**Review by alexarice — COMMENTED** [review]
This is looking great. Thanks for fixing this.

**Comment by alexarice** [comment]
> 
> The one thing I'm currently unsure of is whether an anchored region is allowed to be the first directive in an optional group. The way I have it now, it's not

It's irrelevant whether the first element is the anchor. The important thing is that the first element is marked as optionally parsable, meaning that it successfully parses the empty string

File: tests/irdl/test_declarative_assembly_format.py (hunk 4 above)
Comment (dk949) [reply tests/irdl/test_declarative_assembly_format.py:2158]: Do you mean something like:

```mlir
test.empty_region_group keyword { }
```

It currently roundtrips to:

```mlir
test.empty_region_group
```

Which I believe is consistent with MLIR

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:2158]: I meant something like `test.empty_region_group keyword { ^bb1: }`

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:2158]: I'm having a play round with MLIR now and am less sure about what the intended behaviour here should be.

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:2158]: What was the example op from openmp that uses this feature?

Comment (dk949) [reply tests/irdl/test_declarative_assembly_format.py:2158]: This roundtrips with no changes. I'm not 100% sure what upstream behaviour of this is, but I expect the same?

File: tests/irdl/test_declarative_assembly_format.py (hunk 2 above)
Comment (dk949) [reply tests/irdl/test_declarative_assembly_format.py]: Removed

File: tests/irdl/test_declarative_assembly_format.py (hunk 3 above)
Comment (dk949) [reply tests/irdl/test_declarative_assembly_format.py]: Done

File: tests/irdl/test_declarative_assembly_format.py (hunk 4 above)
Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:2158]: Seems hard to tell what the upstream behaviour is, as most things require a terminator.

Comment (dk949) [reply tests/irdl/test_declarative_assembly_format.py:2158]: Yup. My expectation is that since `keyword {}` becomes nothing, `keyword {^0:}` should retain the empty block, but it's just a guess.

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:2158]: Lets go with this for now and if we find evidence otherwise we can change it (shouldn't effect the ops you care about I think)

Comment (dk949) [reply tests/irdl/test_declarative_assembly_format.py:2158]: > What was the example op from openmp that uses this feature?

`omp.private` uses this:

```mlir
"builtin.module"() ({
  "omp.private"() <{data_sharing_type = #omp<data_sharing_type{type = private}>, sym_name = "p1", type = i32}> ({
  ^bb0(%arg0: i32):
    %0 = "arith.constant"() <{value = 0 : i32}> : () -> i32
    "omp.yield"(%0) : (i32) -> ()
  }, {
  }, {
  }) : () -> ()
}) : () -> ()
```

Becomes:

```mlir
module {
  omp.private {type = private} @p1 : i32 alloc {
  ^bb0(%arg0: i32):
    %c0_i32 = arith.constant 0 : i32
    omp.yield(%c0_i32 : i32)
  }
}
```
